src/x86_64/x86_64.h

```diff
@@ -66,9 +68,94 @@ static inline void disable_interrupts()
 
 typedef u64 *context;
 
-extern context running_frame;
+context allocate_frame(heap h);
+
+typedef struct cpuinfo {
+    /* For accessing cpuinfo via %gs:0; must be first */
+    void * self;
+
+    /* This points to the frame of the current, running context. Entry
+       points expect this to be the second (+8) field here. */
+    context running_frame;
+
+    /* syscall_enter switches to this stack before calling syscall. It
+       must be the third field (+16). */
+    void * syscall_stack;
+
+    /* common_handler switches to this stack when calling process_bhqueue */
+    void * bh_stack;
+
+    /* The default frame for when we're not in a thread or bh context. */
+    context misc_frame;
+
+    /* for bh processing, where page faults must be supported */
```

Comment:
oh...I see, I found this all a bit confusing....but if bh context has to be able to take page faults then I'm totally lost. ... this is for mmap file? no, that can't be it.

Reply:
The bhqueue processing can and does take page faults. Imagine lwip processing invokes a callback which wakes a socket, causing a blocked read to fill data in a user buffer over a non-present anonymous page.

But it is confusing, and again the stack of frames approach could clean this up.